src/library/scala/collection/immutable/Stream.scala

```diff
@@ -1293,6 +1244,23 @@ object Stream extends SeqFactory[Stream] {
   private[immutable] def collectedTail[A, B, That](head: B, stream: Stream[A], pf: PartialFunction[A, B], bf: CanBuildFrom[Stream[A], B, That]) = {
     cons(head, stream.tail.collect(pf)(bf).asInstanceOf[Stream[B]])
   }
-}
-
 
+  /** An implementation of `FilterMonadic` allowing GC of the `Stream` as it is processed.
+    *
+    * Because this is not an inner class of `Stream`, it is now possible to offer explicit
+    * 'null after first use' semantics for the reference to the head of the original
+    * `Stream`, allowing GC to collect while the tail is still processing (see SI-8990).
+    */
```

Comment:
@retronym Yup, that's it!  Not sure whether it should be try/finally or a temporary assignment to a val, though.  Depends on whether you think a second attempt at filtering could succeed.  Could argue it either way.